### PR TITLE
machines: validate if virtual network uses existing IP

### PR DESCRIPTION
**Pull request by skobyda**

### CODE REVIEW

**Review by KKoukiou — COMMENTED** [review]
Tests are failing: https://github.com/cockpit-project/cockpit/pull/14952/commits/86943d524ffb7a4d90c9e264fc0d0b3a88ea56af

Generally I am not sure if we would like to introduce such a check here, especially when that is not straightforward. If libvirt does not throw an error, we should not throw as well.
And if this is something that should be prevented, it's libvirt who should throw the error when defining the network, not us.

** The reason that I am so much hesitant is that it's not clear to me why this check is not done from libvirt, is there any cornern scenario when this could be allowed? Maybe if not both networks are running it's totally fine to have them both defined like this.

**Comment by skobyda** [comment]
> Tests are failing: [86943d5](https://github.com/cockpit-project/cockpit/commit/86943d524ffb7a4d90c9e264fc0d0b3a88ea56af)
> 
> Generally I am not sure if we would like to introduce such a check here, especially when that is not straightforward. If libvirt does not throw an error, we should not throw as well.
> And if this is something that should be prevented, it's libvirt who should throw the error when defining the network, not us.
> 
> ** The reason that I am so much hesitant is that it's not clear to me why this check is not done from libvirt, is there any cornern scenario when this could be allowed? Maybe if not both networks are running it's totally fine to have them both defined like this.

It throws an error. But the error says: "The network is already used by an interface [interface_name]", which is not very useful. It leads user to believe that there is a problem with an interface, and not with network (network address in specific). So I think cockpit could be smarter, user friendly and point out what's the problem specificcally, instead of just showing this confusing libvirt error message.

**Comment by KKoukiou** [comment]
> > Tests are failing: [86943d5](https://github.com/cockpit-project/cockpit/commit/86943d524ffb7a4d90c9e264fc0d0b3a88ea56af)
> > Generally I am not sure if we would like to introduce such a check here, especially when that is not straightforward. If libvirt does not throw an error, we should not throw as well.
> > And if this is something that should be prevented, it's libvirt who should throw the error when defining the network, not us.
> > ** The reason that I am so much hesitant is that it's not clear to me why this check is not done from libvirt, is there any cornern scenario when this could be allowed? Maybe if not both networks are running it's totally fine to have them both defined like this.
> 
> Libvirt does throw an error. But the error says: "The network is already used by an interface [interface_name]", which is not very useful. It leads user to believe that there is a problem with an interface, and not with network (network address in specific) (see #14942). So I think cockpit could be smarter, user friendly and point out what's the problem specificcally, instead of just showing this confusing libvirt error message.

I would saw a warning, and allow the user to create it if the other network is defined but not running. If the other network is defined this should start just fine, right?

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 1]
```diff
@@ -43,17 +43,25 @@ const ConnectionRow = ({ connectionName }) => {
...
-        else if (!utils.validateIpv4(dialogValues.ipv4))
+        } else if (!utils.validateIpv4(dialogValues.ipv4)) {
             validationFailed.ipv4 = _("Invalid IPv4 address");
+        } else {
+            const net = networks.some(net => net.ip[0] &&
+                                      net.ip[0].family === "ipv4" &&
+                                      net.ip[0].netmask === dialogValues.netmask &&
+                                      utils.isIpv4InNetwork(dialogValues.ipv4, dialogValues.netmask, net.ip[0].netmasknetwork.ipv4.address));
+            if (net)
+                validationFailed.ipv4 = cockpit.format(_("IPv4 address is already used by network '$0'"), net.name);
```
Comment (skobyda) [inline pkg/machines/components/networks/createNetworkDialog.jsx:63]: @KKoukiou Yes, the if the other network is not running, they should be allowed to create a network with same address. So I will add this a condition that only if network is active the error message gets rendered.

**Review by KKoukiou — DISMISSED** [review]

**Comment by martinpitt** [comment]
Converting to draft. Please set back to "Ready for review" once the tests got fixed. Thanks!

**Comment by KKoukiou** [comment]
@skobyda you need to move this to the new repository. Please reopen there.

**Pull request closed without merging**